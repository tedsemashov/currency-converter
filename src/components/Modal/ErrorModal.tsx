import { useEffect } from 'react';
import { ModalContainer, ModalText } from './ErrorModalStyles';
import { useAppDispatch } from '../../app/hooks';
import { setError } from './errorSlice';

export const DURATION = 5000;

type ErrorModalProps = {
  text?: string;
};

const ErrorModal = ({ text = '' }: ErrorModalProps) => {
  const dispatch = useAppDispatch();

  useEffect(() => {
    const visibilityTimer = setTimeout(
      () => dispatch(setError({ isError: false })),
      DURATION
    );

    return () => {
      clearTimeout(visibilityTimer);
    };
  });

  return (
    <ModalContainer>
      <ModalText data-testid="modal-error">
        Error: {text}. Try to generate again.
      </ModalText>
    </ModalContainer>
  );
};

export default ErrorModal;
