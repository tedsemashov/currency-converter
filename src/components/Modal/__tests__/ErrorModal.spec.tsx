import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import ErrorModal from '../ErrorModal';
import { render, screen, act } from '@testing-library/react';
import { DURATION } from '../ErrorModal';
import errorReducer, { setError } from '../errorSlice';

const store = createStore(errorReducer);

jest.useFakeTimers();

describe('ErrorModal component', () => {
  it('should return error message', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <ErrorModal text="API server error" />
      </Provider>
    );

    expect(screen.getByTestId('modal-error')).toBeInTheDocument();
    expect(
      screen.getByText('Error: API server error. Try to generate again.')
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should dispatch setError action', () => {
    render(
      <Provider store={store}>
        <ErrorModal text="API server error" />
      </Provider>
    );

    store.dispatch(setError({ isError: true, message: '' }));

    act(() => {
      jest.advanceTimersByTime(DURATION);
    });

    const { isError, message } = store.getState();
    expect(isError).toBe(false);
    expect(message).toBe('');
  });
});
