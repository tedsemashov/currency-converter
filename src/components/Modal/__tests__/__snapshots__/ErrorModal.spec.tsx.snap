// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorModal component should return error message 1`] = `
<DocumentFragment>
  <div
    class="sc-eLqIir Bstic"
  >
    <p
      class="sc-HzRMQ lhygmB"
      data-testid="modal-error"
    >
      Error: API server error. Try to generate again.
    </p>
  </div>
</DocumentFragment>
`;
